### ISSUE TITLE
PERF: Poisson (~28x) and Gamma (~7x) 1-D s(x) fits much slower than mgcv at equal accuracy

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#1690]
## Summary

A plain 1-D smooth under **Poisson** and **Gamma** families fits much slower than the equivalent `mgcv` model at **identical predictive accuracy**:

- Poisson `y ~ s(x)`, n=600: **~28-35x slower** end-to-end, RMSE ratio ≈ 1.0
- Gamma `y ~ s(x)`, n=600: **~6.5-9x slower** end-to-end, RMSE ratio ≈ 1.0

These are not the binomial path (#1575) nor the Gaussian path — they're the count/positive-continuous GLM families on the most basic single-smooth model.

Versions: `gamfit==0.1.236`, `mgcv 1.9.4`. Both `method="REML"` + double penalty (`select=TRUE`). Fit/predict timed separately; gamfit's stderr logging is redirected to `/dev/null` inside the timed region (best case for gamfit).

## Self-contained repro (numpy + gamfit, mgcv via Rscript)

```python
import os, sys, time, json, tempfile, subprocess, contextlib, warnings
warnings.filterwarnings("ignore")
import numpy as np, pandas as pd, gamfit
fam = sys.argv[1] if len(sys.argv) > 1 else "poisson"   # "poisson" or "gamma"

@contextlib.contextmanager
def quiet():
    dn=os.open(os.devnull,os.O_WRONLY); old=os.dup(2); os.dup2(dn,2); os.close(dn)
    try: yield
    finally: os.dup2(old,2); os.close(old)

rng=np.random.default_rng(0); n=600; x=rng.uniform(0,1,n)
mu=np.exp(1.0+np.sin(5*x))
y = rng.poisson(mu).astype(float) if fam=="poisson" else rng.gamma(shape=4.0, scale=mu/4.0)
xt=np.linspace(0,1,400); mut=np.exp(1.0+np.sin(5*xt))
tr=pd.DataFrame({"x":x,"y":y}); te=pd.DataFrame({"x":xt})

bf=bp=1e9
for _ in range(3):
    with quiet():
        t0=time.perf_counter(); m=gamfit.fit(tr,"y ~ smooth(x)",family=fam); t1=time.perf_counter()
        gpr=np.asarray(m.predict(te)).ravel(); t2=time.perf_counter()
    bf=min(bf,t1-t0); bp=min(bp,t2-t1)

R=r'''suppressPackageStartupMessages({library(mgcv);library(jsonlite)})
cfg<-fromJSON(commandArgs(trailingOnly=TRUE)[1]); tr<-read.csv(cfg$train); te<-read.csv(cfg$test)
fm<-if(cfg$fam=="poisson") poisson() else Gamma(link="log")
.t0<-proc.time()[["elapsed"]]; fit<-gam(y~s(x,bs="ps",k=12),family=fm,data=tr,method="REML",select=TRUE); fs<-proc.time()[["elapsed"]]-.t0
.t1<-proc.time()[["elapsed"]]; pr<-as.numeric(predict(fit,newdata=te,type="response")); ps<-proc.time()[["elapsed"]]-.t1
write.csv(data.frame(mean=pr),cfg$out,row.names=FALSE); writeLines(toJSON(list(fit_s=fs,predict_s=ps),auto_unbox=TRUE),cfg$timing)'''
with tempfile.TemporaryDirectory() as td:
    rp=td+"/m.R"; open(rp,"w").write(R)
    trp,tep,op,tp,cp=[td+"/"+a for a in("tr.csv","te.csv","o.csv","t.json","c.json")]
    tr.to_csv(trp,index=False); te.to_csv(tep,index=False)
    json.dump({"train":trp,"test":tep,"out":op,"timing":tp,"fam":fam},open(cp,"w"))
    subprocess.run(["Rscript",rp,cp],check=True)
    t=json.load(open(tp)); mpr=pd.read_csv(op)["mean"].values
rmse=lambda a: float(np.sqrt(np.mean((a-mut)**2)))
print("%s: gamfit %.3fs  mgcv %.3fs  -> %.1fx | rmse %.4f vs %.4f"%(
      fam, bf+bp, t["fit_s"]+t["predict_s"], (bf+bp)/(t["fit_s"]+t["predict_s"]), rmse(gpr), rmse(mpr)))
```

Output on this machine:

```
poisson: gamfit 1.846s  mgcv 0.066s  -> 28.0x | rmse 0.1863 vs 0.1877
gamma:   gamfit 0.730s  mgcv 0.113s  ->  6.5x | rmse 0.2796 vs 0.2720
```

## Multi-seed (stable, not noise)

| family | seeds | gamfit/mgcv total time | RMSE ratio (gamfit/mgcv) |
|---|---|---|---|
| poisson | 0,1 | 28x, 35x | 0.99, 1.03 |
| gamma | 0,1,2,3 | 6.5x, 7.9x, 7.4x, 9.2x | 1.03, 1.00, 0.99, 1.05 |

Accuracy is identical to within noise in every run; gamfit just takes 6-35x longer for the same single-smooth fit. (The Gaussian and Matérn/GP analogues are filed separately; this one is specifically the Poisson/Gamma GLM path.)

**Comment by SauersML** [comment]
**Fleet agent status (from Discussions):** Agent `gam-issue-1690` → PR #1730 (squash-merged to main as 14c503e46). Fixed the actionable Gamma bug: the Gamma `s(x,k=12)` n=600 REML fit reaches the genuine optimum (ρ≈[-1.367,7.533], score 329.259, edf 6.70) but reported `outer_converged=false`. The Gamma log-link profiles the dispersion shape into the inner working weight, so at a weakly-identified flat ρ-valley the inner P-IRLS fixed point and analytic ρ-gradient are warm-start sensitive — the in-loop cost-stall guard sampled `|g|=0.3646` (just above the 0.330 score-relative bound) and halted non-converged; the authoritative gradient at the shipped refined-dispersion β is `|g|=0.2297 < 0.330`, so it IS stationary (Poisson canonical link converges at ~1e-6 and never hits this). That false non-convergence also fired wasted deterministic-replay ARC retries.

Two-part fix: (1) `optimizer.rs` reconciles `outer_converged` against the authoritative `compute_gradient` at the shipped β, gated strictly on the `CostStallFlatValley` stop reason (so the #1426 stuck-overfit `|g|≈11` never clears it); (2) `run.rs` treats `CostStallFlatValley` like `RejectFloor` in the ARC retry gate (drops inner_pirls_solves 146→128). Bit-identical optimum. Regression test `poisson_gamma_outer_work_1690.rs` asserts a stationarity certificate, not a flag flip.

**Out of scope (→ #1033):** the dominant family-agnostic cost — the seed-grid prepass + multistart's ~80 full-n inner solves for a single smooth (mgcv ~8) — is the larger share of the 7-28× and is load-bearing (#1266/#1548/#1464/#1426), deferred.
